cluster.go

```diff
@@ -605,6 +616,18 @@ func (c *clusterClient) doresultfn(ctx context.Context, results *redisresults, r
 				if !c.retry || !cm.IsReadOnly() {
 					continue
 				}
+
+				if !isWaitingForRetry {
+					shouldRetry := c.retryHandler.WaitUntilNextRetry(
+						ctx, attempts, resp.Error(),
+					)
+
+					if !shouldRetry {
+						continue
+					}
+
+					isWaitingForRetry = true
+				}
```

Comment:
Why do you put the waiting here? Instead of putting it at: https://github.com/redis/rueidis/blob/fc348f545e5ac0315ef9350fa7375bda92b8877e/cluster.go#L679-L681

Reply:
If we sleep not here, code is quite complicated. Because we have to distinguish retry or ask/moved in the L679. if retry, should wait. if not retry immediately.

Reply:
Even though there is randomness, i think waiting time for each goroutines, converges similarly. If you want exactly once to wait, i will change it.

Reply:
Thanks for catching this detail I overlooked.

I think a better way to handle this is if any move/ask redirection happens, we don't back off for a retry.

Therefore, how about we add a new counter field to the `retries` struct? So that we can do this:

```go
	if len(retries.m) != 0 {
		if retries.redirects != 0 {
			retries.redirects = 0
			goto retry
		}
		if c.retryHandler.WaitUntilNextRetry(...) {
			goto retry
		}
	}
```

Reply:
We can increment the counter after the lock:
<img width="747" alt="image" src="https://github.com/user-attachments/assets/0cd977b2-4268-4cab-8533-b37ef38672bf">

Reply:
[2df785e](https://github.com/redis/rueidis/pull/643/commits/2df785eacfc7bd765452531bb96d763e21c0dcea)

We support configurable retry or not so should give error to `RetryDelay`. Because of this i use channel to gather errors.

We can't use just int. because it is not thread-safe. I prefer `atomic` in this case, because the type itself guarantees   thread-safe. Even though  `doretry` and `doretrycache` use mutex inside, but it is implicit to me to be thread-safe when load int value.

Reply:
Hi @proost,

I think a `retryErrCh := make(chan error, len(multi))` is too over kill. Let me find a better way for you later.

> but it is implicit to me to be thread-safe when load int value.

It is actually thread-safe, because we read the field after a `wg.Wait()`.

Reply:
Atomic is ok too, but please just use `uint32` which is smaller and has no alignment issue.

Reply:
Hi @proost,

I think a better alternative for `retryErrCh` is calling `RetryDelay` directly in the `doresultfn` function and we store its maximum into the `retries` struct. Could be something like this:

<img width="1358" alt="image" src="https://github.com/user-attachments/assets/f80ac8da-9657-41e9-bb9f-ad2b9728d790">

Reply:
> It is actually thread-safe, because we read the field after a wg.Wait().

Oh, i missed waiting using `sync.WaitGroup`. 

change to use `retries.RetryDelay`.
[f04b433](https://github.com/redis/rueidis/pull/643/commits/f04b433a99cf7b644720617ea5e4c5457ad9882b)